### DISCUSSION TITLE
What is the best approach to using Phlex for a mail view?

### DISCUSSION THREAD

**Discussion by johnofsydney** [yippee-fun/phlex#557] [Help] [answered]
what is the best approach to using Phlex for a mail view?
I made it work fine by simply adding a `require_relative` of my flex view in the existing erb mailer file, and then render th Phlex view into that existing erb mailer. It works, but it feels hacky. Is there a better way?

```rb
# app/views/courier_mailer/approval.html.erb

<% require_relative '../courier_mailer/phlex/approval' %>

<%= render Phlex::Approval.new(courier: @courier) %>

```

**Comment by willcosgrove** [comment, marked as answer]
Hi John! 👋

I don't have first hand experience using Phlex views for mailers, but I also suspect we can find a way that is a bit more straightforward than rendering it inside an ERB template.

It looks like the first problem to solve might be your autoload paths. Judging by the fact that you need the `require_relative`, I'm guessing you don't have your autoload paths configured properly. Or possibly your view isn't named what the autoloader expects it to be named.

The phlex-rails install script adds `app/views`, `app/views/components`, and `app/views/layouts` to the autoload path. For the purposes of this we can get by with just `app/views`, which we could add like this to `application.rb`:

```ruby
module MyApp
  class Application < Rails::Application
    # There's probably other stuff in here, but we can just add this
    # to the bottom of the Application definition
    config.autoload_paths << Rails.root / "app" / "views"
  end
end
```

Now the autoloader will look in your `app/views` directory when looking for constants. Now based on where your file is currently located (`app/views/courier_mailer/phlex/approval.rb`) the constant in that file should look like this:

```ruby
class CourierMailer
  module Phlex
    class Approval < Phlex::HTML
      # ...
    end
  end
end
```

Now inside your `CourierMailer` you should be able to reference `Phlex::Approval` and it will find and load the appropriate file.

Now we need to get it to render! I looked up how to render non-template content into a message, and it looks like something like this should work:

```ruby
# courier_mailer.rb
def approval
  mail(to: "whoever@example.com", subject: "Approved!") do |format|
    format.html { render Phlex::Approval.new }
  end
end
```

The key is to pass a block to `mail` and that block will collect responders for different formats just like `respond_to` does in a controller. Inside the responder block for any given format, you can use `render` to render your Phlex view.

Give that a shot and let me know if you run into any issues!

**Comment by johnofsydney** [comment]
Brilliant! Thanks Will!

so, the generator had added the autoload_paths OK. Turns out I probably didn't need the `require_relative` I guess.

rendering the Phlex view from the courier_mailer worked exactly as you described, now I can get rid of the erb file completely.


Thanks very much.
